[Wiki] Fetch latest CARML updates for internalized libraries (#1587)
Azure/ResourceModules: docs/wiki/Fetching latest changes - Scenario 1 Module library only.md

> In this scenario you have onboarded only the module library as described in [Getting started Scenario 1- Consume library](./Getting%20started%20-%20Scenario%201%20Consume%20library) and therefore you fetch lastest changes for the modules.

Reviewer: ```suggestion
In this scenario you have onboarded only the module library as described in [Getting started Scenario 1- Consume library](./Getting%20started%20-%20Scenario%201%20Consume%20library) and you would therefore need to fetch latest changes for the modules.
```

Modules can be stored in an accessible location like local, Template Specs, Bicep Registry or ADO universal packages.

The update process is the following:

### _Navigation_

- [1. Backup your local copy of the library](#1-backup-your-local-copy-of-the-library)
- [2. Download the library](#2-download-the-library)
- [3. (Optional) Convert library to ARM](#3-optional-convert-library-to-arm)
- [4. (Optional) Customize modules](#4-optional-customize-modules)
- [5. Test and publish modules](#5-test-and-publish-modules)

> # 1. Backup your local copy of the library

Reviewer: ```suggestion

# 1. Backup your local copy of the library
```

# 4. (Optional) Customize modules

There are different options for library's customization:
- [Option 1 - Library improvement](./Fetching%20latest%20updates#option-1---library-improvement)
- [Option 2 - Specific requirements](./Fetching%20latest%20updates#option-2---specific-requirements)

The recommendation is to [contribute](./Contribution%20guide) to the public CARML repository so that your updates can improve the public library. In this way your changes will be available on the public library when fetching updates, modules will be already tested with your changes and you won't need to take care of customization on each update.

In some cases, you might need to add to the library company/organization' specifics, that are not sharable with the public CARML repository.

In this scenario every time you'll fetch updates from the public CARML repository merge conflicts are expected. You'll have to compare the new public code with your customized one and re-apply your customizations to the updated code.

This process can be automated, by script or CI, if customization tasks are repeatable.

> # 5. Test and publish modules

Reviewer: ```suggestion

# 5. Test and publish modules
```

In this scenario you have onboarded only the module library as described in [Getting started Scenario 1- Consume library](./Getting%20started%20-%20Scenario%201%20Consume%20library) and therefore you fetch lastest changes for the modules.

Modules can be stored in an accessible location like local, Template Specs, Bicep Registry or ADO universal packages.

The update process is the following:

### _Navigation_

- [1. Backup your local copy of the library](#1-backup-your-local-copy-of-the-library)
- [2. Download the library](#2-download-the-library)
- [3. (Optional) Convert library to ARM](#3-optional-convert-library-to-arm)
- [4. (Optional) Customize modules](#4-optional-customize-modules)
- [5. Test and publish modules](#5-test-and-publish-modules)

> # 1. Backup your local copy of the library

Reviewer: ```suggestion

# 1. Backup your local copy of the library
```

> <details>
> <summary>Rename your local repository</summary>
> 
> Assuming the local repository location is `'D:\ResourcesModules'` rename it in `'D:\ResourcesModules_Backup'`.
> 
> The backup folder can be used to compare customized modules with the ones coming from the latest version and implement the required changes.
> 
> This can be done, for example, by the `'Compare selected'` [function](https://vscode.one/diff-vscode/) of Visual Studio Code
> 
> </details>

Reviewer: ```suggestion
Rename your local repository. Assuming the local repository location is `'D:\ResourcesModules'` rename it in `'D:\ResourcesModules_Backup'`.

The backup folder can be used to compare customized modules with the ones coming from the latest version and implement the required changes.

This can be done, for example, by the `'Compare selected'` [function](https://vscode.one/diff-vscode/) of Visual Studio Code

```

# 2. Download the library

To download the updated library follow the best option for your scenario, as explained in [Download the library](./Getting%20started%20-%20Scenario%202%20Consume%20library#1-download-the-library)

# 3. (Optional) Convert library to ARM

In case you aren't using Bicep you need to follow this procedure: [(Optional) Convert library to ARM](./Getting%20started%20-%20Scenario%202%20Consume%20library#2-optional-convert-library-to-arm)

> # 4. (Optional) Customize modules

Reviewer: ```suggestion

# 4. (Optional) Customize modules
```

There are different options for library's customization:
- [Option 1 - Library improvement](./Fetching%20latest%20updates#option-1---library-improvement)
- [Option 2 - Specific requirements](./Fetching%20latest%20updates#option-2---specific-requirements)

The recommendation is to [contribute](./Contribution%20guide) to the public CARML repository so that your updates can improve the public library. In this way your changes will be available on the public library when fetching updates, modules will be already tested with your changes and you won't need to take care of customization on each update.

In some cases, you might need to add to the library company/organization' specifics, that are not sharable with the public CARML repository.

In this scenario every time you'll fetch updates from the public CARML repository merge conflicts are expected. You'll have to compare the new public code with your customized one and re-apply your customizations to the updated code.

This process can be automated, by script or CI, if customization tasks are repeatable.

> # 5. Test and publish modules

Reviewer: ```suggestion

# 5. Test and publish modules
```

> When the CARML library is consumed with the [*Template Orchestration*](./Solution%20creation#template-orchestration) approach modules need to be stored in an available location, where the Azure Resource Manager (ARM) can access them. This can be achieved by storing the modules templates in an accessible location like local, Template Specs or the Bicep Registry.

Reviewer: As mentioned for Scenario 2, I suggest not to map scenarios 1 and 2 with orchestrations and remove this reference to avoid confusion. Modules need to be stored in an available location also if leveraging pipeline orchestration.

> When the CARML library is consumed with the [*Template Orchestration*](./Solution%20creation.md#template-orchestration) approach modules need to be stored in an available location, where the Azure Resource Manager (ARM) can access them. This can be achieved by storing the modules templates in an accessible location like local, Template Specs or the Bicep Registry.

Reviewer 2: ```suggestion
When the CARML library is consumed with the [*Template Orchestration*](./Solution%20creation#template-orchestration) approach modules need to be stored in an available location, where the Azure Resource Manager (ARM) can access them. This can be achieved by storing the modules templates in an accessible location like local, Template Specs or the Bicep Registry.
```

---
### _Navigation_

- [Local Repository](#local-repository)
- [Template Specs](#template-specs)

> - [Bicep Registry](#bicep-registry)

Reviewer: ADO universal packages should also be added as an option.

---

# Local Repository
In case you consume the module library with a local repository, the update procedure has the following steps:

> 1. Backup your local repository to maintain a backup copy

Reviewer: It looks like all the steps besides the last one (publishing) are the same for all options, i.e. backup, download, conversion and customization. I'd suggest to describe them just once for maintainability and then differentiate the last step (publishing) based on the approach.

Author: grouped by procedure instead of repository

    <details>
    <summary>Rename your local repository</summary>

    Assuming the local repository location is `'D:\ResourcesModules'` rename it in `'D:\ResourcesModules_Backup'`.
    The backup folder can be used to compare customized modules with the ones coming from the latest version and implement the required changes.
    This can be done, for example, by the `'Compare selected'` [function](https://vscode.one/diff-vscode/) of Visual Studio Code
    </details>

> 1. [Download the library](./Getting%20started%20-%20Scenario%202%20Consume%20library.md#1-download-the-library)

Reviewer 2: ```suggestion
1. [Download the library](./Getting%20started%20-%20Scenario%202%20Consume%20library#1-download-the-library)
```

> 1. [Download the library](./Getting%20started%20-%20Scenario%202%20Consume%20library.md#1-download-the-library)

Reviewer 2: ```suggestion
1. [Download the library](./Getting%20started%20-%20Scenario%202%20Consume%20library#1-download-the-library)
```

Author: all .md removed

> 1. [(Optional) Convert library to ARM](./Getting%20started%20-%20Scenario%202%20Consume%20library.md#2-optional-convert-library-to-arm)

Reviewer 2: ```suggestion
1. [(Optional) Convert library to ARM](./Getting%20started%20-%20Scenario%202%20Consume%20library#2-optional-convert-library-to-arm)
```

> If you have customized modules, you'll need to apply customizations on the new updated modules.

Reviewer: This is the main reason for the `fetching latest updates` section, hence I'd suggest to elaborate more on this point. e.g. highlight conflicts are expected and can be minimized by contributing back to the main repository if changes are useful for the public.

Author: added customization explanation

# Template Specs
When the modules are stored in Template Specs a versioning feature with access to old versions is available, but a backup of the local folder can be useful if you have customized modules, as you'll need to apply the customization to the new downloaded version of the modules. The update steps are the following:
1. Backup your local repository to maintain a backup copy: you can rename the local repository as showed in the above section.

> 1. [Download the library](./Getting%20started%20-%20Scenario%202%20Consume%20library.md#1-download-the-library)

Reviewer 2: ```suggestion
1. [Download the library](./Getting%20started%20-%20Scenario%202%20Consume%20library#1-download-the-library)
```

> 1. [(Optional) Convert library to ARM](./Getting%20started%20-%20Scenario%202%20Consume%20library.md#2-optional-convert-library-to-arm)

Reviewer 2: ```suggestion
1. [(Optional) Convert library to ARM](./Getting%20started%20-%20Scenario%202%20Consume%20library#2-optional-convert-library-to-arm)
```

1. Apply customizations to the new version of modules if needed
1. Publish modules to Template Spec
    <details>
    <summary>Modules publishing in Template Spec</summary>

>     The preferred method to publish modules to template-specs is to leverage CARML ready [CI environment](./The%20CI%20environment.md), however there maybe specific requirements for which this option is not applicable. As an alternative, the same [Publish-ModuleToTemplateSpec.ps1](https://github.com/Azure/ResourceModules/blob/main/utilities/pipelines/resourcePublish/Publish-ModuleToPrivateBicepRegistry.ps1) script leveraged by the publishing step of the CI environment pipeline can be executed locally.

Reviewer 2: ```suggestion
    The preferred method to publish modules to template-specs is to leverage CARML ready [CI environment](./The%20CI%20environment), however there maybe specific requirements for which this option is not applicable. As an alternative, the same [Publish-ModuleToTemplateSpec.ps1](https://github.com/Azure/ResourceModules/blob/main/utilities/pipelines/resourcePublish/Publish-ModuleToPrivateBicepRegistry.ps1) script leveraged by the publishing step of the CI environment pipeline can be executed locally.
```

        - TemplateFilePath = the absolute path of the module to be published
        - ModuleVersion = the version of the module
        - TemplateSpecsRgName = the resource group that will contain the Template Spec
        - TemplateSpecsRgLocation = the location of the Template Spec
        - TemplateSpecsDescription = The description of the Template Spec

       To publish the Keyvault module with version 0.4.740 on a Template Spec that will be created in the resource group 'artifact-rg' you can execute the following example:
         ```PowerShell
        Publish-ModuleToTemplateSpec -TemplateFilePath "D:\ResourcesModules\arm\Microsoft.KeyVault\vaults\deploy.bicep" -ModuleVersion "0.4.740" -TemplateSpecsRgName 'artifact-rg'  -TemplateSpecsRgLocation 'West Europe' -TemplateSpecsDescription 'CARML KV Template Spec'
        ```
        As the modules to be published are more than one a script that calls the `'Publish-ModuleToTemplateSpec'` function for each of the modules can be created.
    </details>
    <p
1. Update your master template in order to use the new version of the published modules.

>     For the [Template Specs' example in Solutions](./Solution%20creation.md#examples) page, supposing you have published version '0.4.740' of modules, you need to replace all the occurences of '0.4.735' with '0.4.740'.

Reviewer 2: ```suggestion
    For the [Template Specs' example in Solutions](./Solution%20creation#examples) page, supposing you have published version '0.4.740' of modules, you need to replace all the occurences of '0.4.735' with '0.4.740'.
```

# Bicep Registry
As the Template Specs scenario, Bicep Registry offers a versioning feature with access to old versions of modules, but a backup of the local folder can be useful if you have customized modules, as you'll need to apply the customization to the new downloaded version of the modules. The update steps are the following:
1. Backup your local repository to maintain a backup copy: you can rename the local repository as showed in the above section.

> 1. [Download the library](./Getting%20started%20-%20Scenario%202%20Consume%20library.md#1-download-the-library)

Reviewer 2: ```suggestion
1. [Download the library](./Getting%20started%20-%20Scenario%202%20Consume%20library#1-download-the-library)
```

> 1. [(Optional) Convert library to ARM](./Getting%20started%20-%20Scenario%202%20Consume%20library.md#2-optional-convert-library-to-arm)

Reviewer 2: ```suggestion
1. [(Optional) Convert library to ARM](./Getting%20started%20-%20Scenario%202%20Consume%20library#2-optional-convert-library-to-arm)
```

1. Apply customizations to the new version of modules if needed
1. Publish modules to Bicep Registry
    <details>
    <summary>Modules publishing in Bicep Registry</summary>

>     The preferred method to publish modules to Bicep Registry is to leverage CARML ready [CI environment](./The%20CI%20environment.md), however there maybe specific requirements for which this option is not applicable. As an alternative, the same [Publish-ModuleToPrivateBicepRegistry.ps1](https://github.com/Azure/ResourceModules/blob/main/utilities/pipelines/resourcePublish/Publish-ModuleToPrivateBicepRegistry.ps1) script leveraged by the publishing step of the CI environment pipeline can be executed locally.

Reviewer 2: ```suggestion
    The preferred method to publish modules to Bicep Registry is to leverage CARML ready [CI environment](./The%20CI%20environment), however there maybe specific requirements for which this option is not applicable. As an alternative, the same [Publish-ModuleToPrivateBicepRegistry.ps1](https://github.com/Azure/ResourceModules/blob/main/utilities/pipelines/resourcePublish/Publish-ModuleToPrivateBicepRegistry.ps1) script leveraged by the publishing step of the CI environment pipeline can be executed locally.
```

Author: Thank you Alex, i didn't know it
